drivers/i2c/i2c_renesas_ra_iic.c

```diff
@@ -0,0 +1,552 @@
+					"(65535) while DTC is used for data transfer.",
+					__func__);
+				break;
+			case FSP_ERR_IN_USE:
+				LOG_ERR("%s: Bus busy condition. Another transfer was in progress.",
+					__func__);
+				break;
+			default:
+				/* Should not reach here. */
+				LOG_ERR("%s: Unknown error. FSP_ERR=%d\n", __func__, fsp_err);
+				break;
+			}
+
+			ret = -EIO;
+			goto RELEASE_BUS;
+		}
+
+		/* Wait for callback to return. */
+		k_sem_take(&data->complete_sem, K_FOREVER);
+
+		/* Handle event msg from callback. */
+		switch (data->event) {
+		case I2C_MASTER_EVENT_ABORTED:
+			LOG_ERR("%s: %s failed.", __func__,
+				(current->flags & I2C_MSG_READ) ? "Read" : "Write");
+			ret = -EIO;
+			goto RELEASE_BUS;
+		case I2C_MASTER_EVENT_RX_COMPLETE:
+			break;
+		case I2C_MASTER_EVENT_TX_COMPLETE:
+			break;
+		default:
+			break;
+		}
+
+		current++;
+		num_msgs--;
+	}
+
+RELEASE_BUS:
+	k_mutex_unlock(&data->bus_mutex);
+
+	return ret;
+}
+
+static void i2c_ra_iic_callback(i2c_master_callback_args_t *p_args)
+{
+	const struct device *dev = p_args->p_context;
+	struct i2c_ra_iic_data *data = dev->data;
+
+	data->event = p_args->event;
+
+	k_sem_give(&data->complete_sem);
+}
+
+static int i2c_ra_iic_init(const struct device *dev)
+{
+	const struct i2c_ra_iic_config *config = dev->config;
+	struct i2c_ra_iic_data *data = (struct i2c_ra_iic_data *)dev->data;
+	fsp_err_t fsp_err = FSP_SUCCESS;
+	int ret = 0;
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+
+	if (ret < 0) {
+		LOG_ERR("%s: pinctrl config failed.", __func__);
+		return ret;
+	}
+
+	k_mutex_init(&data->bus_mutex);
+	k_sem_init(&data->complete_sem, 0, 1);
+
+	switch (data->fsp_config.rate) {
+	case I2C_MASTER_RATE_STANDARD:
+	case I2C_MASTER_RATE_FAST:
+	case I2C_MASTER_RATE_FASTPLUS:
+		calc_iic_master_clock_setting(dev, data->fsp_config.rate,
+					      &data->iic_master_ext_cfg.clock_settings);
+		data->iic_master_ext_cfg.timeout_mode = IIC_MASTER_TIMEOUT_MODE_SHORT;
+		data->iic_master_ext_cfg.timeout_scl_low = IIC_MASTER_TIMEOUT_SCL_LOW_ENABLED;
+
+		data->fsp_config.p_extend = &data->iic_master_ext_cfg;
+		break;
+	default:
+		LOG_ERR("%s: Invalid I2C speed rate: %d", __func__, data->fsp_config.rate);
+		return -ENOTSUP;
+	}
+
+	fsp_err = R_IIC_MASTER_Open(&data->ctrl, &data->fsp_config);
+	__ASSERT(fsp_err == FSP_SUCCESS, "%s: Open iic master failed. FSP_ERR=%d", __func__,
+		 fsp_err);
+
+	config->irq_config_func(dev);
+
+	return 0;
+}
+
+static void calc_iic_master_bitrate(const struct i2c_ra_iic_config *config, uint32_t total_brl_brh,
+				    uint32_t brh, uint32_t divider,
+				    struct ra_iic_master_bitrate *result)
+{
+	const uint32_t noise_filter_stage = config->noise_filter_stage;
+	const double rise_time_ns = config->rise_time_ns / RA_IIC_MASTER_DIV_TIME_NS;
+	const double fall_time_ns = config->fall_time_ns / RA_IIC_MASTER_DIV_TIME_NS;
```

Comment:
Could we move those divisions in `i2c_ra_iic_config_##index` definition so it will happen in compile time?